pkg/federate/kubefed/resource.go

```diff
@@ -41,16 +46,44 @@ func (f *federator) Distribute(resource runtime.Object, clusterNames ...string)
 	}
 
 	// Update first, as it will probably be the most common operation
-	err = f.kubeFedClient.Update(context.TODO(), fedResource)
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		existing := &unstructured.Unstructured{}
+		existing.SetGroupVersionKind(fedResource.GroupVersionKind())
+
+		getErr := f.kubeFedClient.Get(context.TODO(), client.ObjectKey{Namespace: fedResource.GetNamespace(), Name: fedResource.GetName()}, existing)
+		if getErr != nil {
+			return getErr
+		}
+
+		// Replace the "spec" field in the existing object with the updated field from "fedResource". We preserve the
+		// existing metadata info, specifically the ResourceVersion which must be set on an update operation.
+		// Note - we set a valid "spec" field earlier in "fedResource" so no need to handle the other return values.
+		spec, _, _ := unstructured.NestedFieldNoCopy(fedResource.Object, ctlutil.SpecField)
+
+		if setErr := unstructured.SetNestedField(existing.Object, spec, ctlutil.SpecField); setErr != nil {
+			return fmt.Errorf(errorSettingFederatedFields, setErr)
+		}
+
+		klog.V(2).Infof("Updating federated resource: %#v", existing)
+
+		return f.kubeFedClient.Update(context.TODO(), existing)
+	})
+
 	if err == nil {
 		return nil
 	}
 
 	if !errors.IsNotFound(err) {
-		return err
+		return fmt.Errorf("error updating federated resoource %#v: %v", fedResource, err)
+	}
+
+	klog.V(2).Infof("Creating federated resource: %#v", fedResource)
+
+	if err = f.kubeFedClient.Create(context.TODO(), fedResource); err != nil {
```

Comment:
Since now you are "getting" before issuing an update you know beforehand if there is a need for creation or not. Thus this flow is not needed anymore. You should refactor for readability.

Reply:
Not sure I follow...